pkg/exchange/bitget/stream.go

```diff
@@ -0,0 +1,167 @@
+package bitget
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/c9s/bbgo/pkg/exchange/bitget/bitgetapi"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+//go:generate callbackgen -type Stream
+type Stream struct {
+	types.StandardStream
+
+	booksEventCallbacks []func(o BookEvent)
+}
+
+func NewStream() *Stream {
+	stream := &Stream{
+		StandardStream: types.NewStandardStream(),
+	}
+
+	stream.SetEndpointCreator(stream.createEndpoint)
+	stream.SetParser(stream.parseWebSocketEvent)
+	stream.SetDispatcher(stream.dispatchEvent)
+	stream.OnConnect(stream.handlerConnect)
+
+	stream.OnBooksEvent(stream.handleBooksEvent)
+	return stream
+}
+
+func (s *Stream) syncSubscriptions(opType WsEventType) error {
+	if opType != WsEventUnsubscribe && opType != WsEventSubscribe {
+		return fmt.Errorf("unexpected subscription type: %v", opType)
+	}
+
+	logger := log.WithField("opType", opType)
+	args := []WebsocketArg{}
+	for _, subscription := range s.Subscriptions {
+		arg, err := s.convertSubscription(subscription)
+		if err != nil {
+			logger.WithError(err).Errorf("convert error, subscription: %+v", subscription)
+			return err
+		}
+
+		args = append(args, arg)
+	}
+
+	logger.Infof("%s channels: %+v", opType, args)
+	if err := s.Conn.WriteJSON(WsOp{
+		Op:   opType,
+		Args: args,
+	}); err != nil {
+		logger.WithError(err).Error("failed to send request")
+		return err
+	}
+
+	return nil
+}
+
+func (s *Stream) Unsubscribe() {
+	// errors are handled in the syncSubscriptions, so they are skipped here.
+	_ = s.syncSubscriptions(WsEventUnsubscribe)
+	s.Resubscribe(func(old []types.Subscription) (new []types.Subscription, err error) {
+		// clear the subscriptions
+		return []types.Subscription{}, nil
+	})
+}
+
+func (s *Stream) createEndpoint(_ context.Context) (string, error) {
+	var url string
+	if s.PublicOnly {
+		url = bitgetapi.PublicWebSocketURL
+	} else {
+		url = bitgetapi.PrivateWebSocketURL
+	}
+	return url, nil
+}
+
+func (s *Stream) dispatchEvent(event interface{}) {
+	switch e := event.(type) {
+	case *WsOpEvent:
+		if err := e.IsValid(); err != nil {
+			log.Errorf("invalid event: %v", err)
+		}
+
+	case *BookEvent:
+		s.EmitBooksEvent(*e)
+	}
+}
+
+func (s *Stream) parseWebSocketEvent(in []byte) (interface{}, error) {
+	var e WsEvent
+
+	err := json.Unmarshal(in, &e)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case e.IsOp():
+		return e.WsOpEvent, nil
+
+	case e.IsChannel():
+		switch e.Arg.Channel {
+		case ChannelOrderBook, ChannelOrderBook5, ChannelOrderBook15:
+			var books BookEvent
+			err = json.Unmarshal(e.WsChannelEvent.Data, &books.Events)
+			if err != nil {
+				return nil, fmt.Errorf("failed to unmarshal data into BookEvent: %+v, err: %w", string(e.WsChannelEvent.Data), err)
```

Comment:
Could we log `e.Arg.Channel` as well? This would allow us to immediately identify which channel is problematic upon encountering the error.